service/history/queues/executable_dlq.go

```diff
@@ -0,0 +1,107 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queues
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/serialization"
+)
+
+type (
+	// ExecutableDLQ is an Executable decorator that will enqueue the task to a DLQ if the underlying Executable fails.
+	// The first call to Execute for which the underlying Executable returns a terminal error will return an
+	// ErrTerminalTaskFailure error and change the behavior of this task to just try to send itself to the DLQ.
+	// Specifically, all subsequent calls to Execute will attempt to send the task to the DLQ. If this fails, the error
+	// will be returned (with the expectation that clients will retry). If it succeeds, no error will be returned and
+	// the task can be acked. When the executable is in the failed state, all calls to HandleErr will just return the
+	// error passed in.
+	// TODO: wrap all executables with this
+	// TODO: add metrics and logging to this
+	ExecutableDLQ struct {
+		Executable
+		dlq DLQ
+		// dlqCause is the original error which caused this task to be sent to the DLQ. It is only set once.
+		dlqCause        error
+		clusterMetadata ClusterMetadata
+	}
+
+	// DLQ is a dead letter queue that can be used to enqueue tasks that fail to be processed.
+	DLQ interface {
+		EnqueueTask(
+			ctx context.Context,
+			request *persistence.EnqueueTaskRequest,
+		) (*persistence.EnqueueTaskResponse, error)
+	}
+
+	// ClusterMetadata is an interface that provides access to cluster metadata.
+	ClusterMetadata interface {
+		GetCurrentClusterName() string
+	}
+)
+
+var (
+	_                      Executable = new(ExecutableDLQ)
+	ErrTerminalTaskFailure            = errors.New("original task failed and this task is now to send the original to the DLQ")
+	ErrSendTaskToDLQ                  = errors.New("failed to send task to DLQ")
+)
+
+// NewExecutableDLQ wraps an Executable to ensure that it is sent to the DLQ if it fails terminally.
+func NewExecutableDLQ(dlq DLQ, executable Executable, clusterMetadata ClusterMetadata) *ExecutableDLQ {
```

Comment:
looks like we just need to pass in the current cluster name?

Reply:
Fixed, yeah that's much simpler.